### PR TITLE
Fix `SegmentedButton` states update logic

### LINKED ISSUE #140746: SegmentedButton displays wrong style on rebuild if disabled
### Steps to reproduce

https://dartpad.dev/?id=fbfa77b9488f667f29c87ab672cdcea4 demonstrates the issue simply.

A two segment `SegmentedButton` widget has a `null` value for `onSelectionChanged` and when the widget is rebuilt when `setState()` is called, the widget changes its style even though no actual state was [changed.[](url)](url)

### Expected results

The `SegmentedButton` widget should maintain its "disabled" style on rebuild.

### Actual results

The `SegmentedButton` widget, on rebuild, has a style that appears it's enabled.

### Code sample

https://dartpad.dev/?id=fbfa77b9488f667f29c87ab672cdcea4

### Screenshots or Video

_No response_

### Logs

_No response_

### Flutter Doctor output

Dartpad reports "Based on Flutter 3.16.5 Dart SDK 3.2.3"

On my local installation, where I first discovered this:

```
$ flutter doctor -v
[✓] Flutter (Channel stable, 3.16.5, on Ubuntu 22.04.3 LTS 6.2.0-39-generic, locale en_US.UTF-8)
    • Flutter version 3.16.5 on channel stable at /home/jpwoodbu/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 78666c8dc5 (10 days ago), 2023-12-19 16:14:14 -0800
    • Engine revision 3f3e560236
    • Dart version 3.2.3
    • DevTools version 2.28.4

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /home/jpwoodbu/Android/Sdk/
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /opt/google/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • Ubuntu clang version 14.0.0-1ubuntu1.1
    • cmake version 3.22.1
    • ninja version 1.10.1
    • pkg-config version 0.29.2

[✓] Android Studio (version 2023.1)
    • Android Studio at /opt/google/android-studio
    • Flutter plugin version 77.1.2
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.85.1)
    • VS Code at /usr/share/code
    • Flutter extension version 3.80.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 22.04.3 LTS 6.2.0-39-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 120.0.6099.129

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

### PR DESCRIPTION
**Pull request by sharabiddin**
This PR fixes: #140746

Desc: Style state update logic fix on SegmentedButton

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

File: packages/flutter/lib/src/material/segmented_button.dart [hunk 1]
```diff
@@ -346,9 +346,11 @@ class SegmentedButtonState<T> extends State<SegmentedButton<T>> {
...
           : null;
       final MaterialStatesController controller = statesControllers.putIfAbsent(segment, () => MaterialStatesController());
-      controller.value = <MaterialState>{
-          if (segmentSelected) MaterialState.selected,
-      };
+      if (segmentSelected) {
+        controller.value.add(MaterialState.selected);
+      } else {
+        controller.value.remove(MaterialState.selected);
+      }
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/segmented_button.dart:353]:
```suggestion
      controller.update(MaterialState.selected, segmentSelected);
```

File: packages/flutter/test/material/segmented_button_test.dart [hunk 2]
```diff
@@ -664,4 +664,32 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
...
+                }),
+          ),
+        ),
+      ),
+    );
+    await tester.pump();
+    final SegmentedButtonState<int> state = tester.state(find.byType(SegmentedButton<int>));
+    const MaterialState disabled = MaterialState.disabled;
+    expect(state.statesControllers.values.first.value.last, disabled);
+  });
```
Comment (TahaTesser) [inline packages/flutter/test/material/segmented_button_test.dart:693]:
```suggestion

  testWidgets('Disabled SegmentedButton has correct states when rebuilding', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: Center(
            child: StatefulBuilder(
              builder: (BuildContext context, StateSetter setState) {
                return Column(
                  children: <Widget>[
                    SegmentedButton<int>(
                      segments: const <ButtonSegment<int>>[
                        ButtonSegment<int>(value: 0, label: Text('foo')),
                      ],
                      selected: const <int>{0},
                    ),
                    ElevatedButton(
                      onPressed: () {
                        setState(() { });
                      },
                      child: const Text('Trigger rebuild'),
                    ),
                  ],
                );
              }
            ),
          ),
        ),
      ),
    );

    final Set<MaterialState> states = <MaterialState>{ MaterialState.selected, MaterialState.disabled };

    // Check the initial states.
    SegmentedButtonState<int> state = tester.state(find.byType(SegmentedButton<int>));
    expect(state.statesControllers.values.first.value, states);

    // Trigger a rebuild.
    await tester.tap(find.byType(ElevatedButton));
    await tester.pumpAndSettle();

    // Check the states after the rebuild.
    state = tester.state(find.byType(SegmentedButton<int>));
    expect(state.statesControllers.values.first.value, states);
  });
```

File: packages/flutter/test/material/segmented_button_test.dart [hunk 3]
```diff
@@ -664,4 +664,32 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
...
+
```
Comment (TahaTesser) [inline packages/flutter/test/material/segmented_button_test.dart:694]:
```suggestion
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]
Thanks for contributing! I've applied some updates to your code.

**Comment by sharabiddin** [comment]
@TahaTesser  Thanks for the changes you suggested, I appreciate it. All resolved

**Review by TahaTesser — CHANGES_REQUESTED** [review]
LGTM!

**Review by TahaTesser — APPROVED** [review]

File: packages/flutter/test/material/segmented_button_test.dart [hunk 4]
```diff
@@ -774,8 +774,48 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
...
+    // Check the states after the rebuild.
+    state = tester.state(find.byType(SegmentedButton<int>));
+    expect(state.statesControllers.values.first.value, states);
+  });
 }
 
 Set<MaterialState> enabled = const <MaterialState>{};
 Set<MaterialState> disabled = const <MaterialState>{ MaterialState.disabled };
 Set<MaterialState> selected = const <MaterialState>{ MaterialState.selected };
+
```
Comment (TahaTesser) [inline packages/flutter/test/material/segmented_button_test.dart:821]:
```suggestion
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: packages/flutter/test/material/segmented_button_test.dart (hunk 4 above)
Comment (sharabiddin) [reply packages/flutter/test/material/segmented_button_test.dart:821]: https://dart.dev/tools/linter-rules/eol_at_end_of_file (sure?) I can remove it if it is okay but against linter

Comment (TahaTesser) [reply packages/flutter/test/material/segmented_button_test.dart:821]: I might be mistaking.  This is the error from the analyzer in the PR

`║    info • Missing a newline at the end of the file • packages/flutter/test/material/segmented_button_test.dart:820:79 • eol_at_end_of_file`

Comment (sharabiddin) [reply packages/flutter/test/material/segmented_button_test.dart]: I was), fixed

**Review by TahaTesser — APPROVED** [review]
LGTM again

**Comment by TahaTesser** [comment]
We will a second review from Hans before we can land this.

File: packages/flutter/test/material/segmented_button_test.dart [hunk 5]
```diff
@@ -304,8 +304,8 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
 
     Finder textHasIcon(String text, IconData icon) {
       return find.descendant(
-        of: find.widgetWithText(Row, text),
-        matching: find.byIcon(icon)
+          of: find.widgetWithText(Row, text),
```
Comment (HansMuller) [inline packages/flutter/test/material/segmented_button_test.dart:307]: This indent isn't needed. LIkewise for lines 308, 344, 345. I'll merge a fix.
```suggestion
        of: find.widgetWithText(Row, text),
```

File: packages/flutter/test/material/segmented_button_test.dart [hunk 6]
```diff
@@ -341,8 +341,8 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
...
```
Comment (HansMuller) [inline packages/flutter/test/material/segmented_button_test.dart:344]:
```suggestion
        of: find.widgetWithText(Row, text),
```

File: packages/flutter/test/material/segmented_button_test.dart [hunk 7]
```diff
@@ -341,8 +341,8 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
...
+          matching: find.byIcon(icon)
```
Comment (HansMuller) [inline packages/flutter/test/material/segmented_button_test.dart:345]:
```suggestion
        matching: find.byIcon(icon)
```

File: packages/flutter/test/material/segmented_button_test.dart [hunk 8]
```diff
@@ -384,8 +384,8 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
 
     Finder rowWithIcons(IconData icon1, IconData icon2) {
       return find.descendant(
-        of: find.widgetWithIcon(Row, icon1),
-        matching: find.byIcon(icon2)
+          of: find.widgetWithIcon(Row, icon1),
```
Comment (HansMuller) [inline packages/flutter/test/material/segmented_button_test.dart:387]:
```suggestion
        of: find.widgetWithIcon(Row, icon1),
```

File: packages/flutter/test/material/segmented_button_test.dart [hunk 9]
```diff
@@ -384,8 +384,8 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
...
+          matching: find.byIcon(icon2)
```
Comment (HansMuller) [inline packages/flutter/test/material/segmented_button_test.dart:388]:
```suggestion
        matching: find.byIcon(icon2)
```

File: packages/flutter/test/material/segmented_button_test.dart [hunk 10]
```diff
@@ -774,6 +774,45 @@ testWidgets('SegmentedButton shows checkboxes for selected segments', (WidgetTes
...
+                        ],
+                        selected: const <int>{0},
+                      ),
+                      ElevatedButton(
+                        onPressed: () => setState(() {}),
+                        child: const Text('Trigger rebuild'),
+                      ),
+                    ],
+                  );
+                }
```
Comment (HansMuller) [inline packages/flutter/test/material/segmented_button_test.dart:799]:
```suggestion
              builder: (BuildContext context, StateSetter setState) {
                return Column(
                  children: <Widget>[
                    SegmentedButton<int>(
                      segments: const <ButtonSegment<int>>[
                        ButtonSegment<int>(value: 0, label: Text('foo')),
                      ],
                      selected: const <int>{0},
                    ),
                    ElevatedButton(
                      onPressed: () => setState(() {}),
                      child: const Text('Trigger rebuild'),
                    ),
                  ],
                );
              },
```

**Review by HansMuller — COMMENTED** [review]
Looks good, just some indentation issues. I'll merge those.

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by sharabiddin** [comment]
Thanks, @HansMuller and @TahaTesser for your efforts. I saw it was merged and you guys made my day

**Pull request merged**